functions/go/annotate-mutations/main.go

```diff
@@ -0,0 +1,159 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt-functions-catalog/functions/go/annotate-mutations/annotatemutations"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/fn/framework/command"
+	"sigs.k8s.io/kustomize/kyaml/kio/kioutil"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+const (
+	mutationCommentIdentifier = "# apply-time-mutation: "
+	annotationKey             = "config.kubernetes.io/apply-time-mutation"
+)
+
+type mutatorAnnotation []annotatemutations.Mutation
+
+type commentProcessor struct {
+	results []framework.ResultItem
+}
+
+type fieldWalker struct {
+	parentProcessor *commentProcessor
+	fileName        string
+	// mutationCount increments as field mutations are found, to ensure unique replacement tokens in multi-comment resources.
+	mutationCount int
+	annotation    mutatorAnnotation
+}
+
+// extractMutationPattern extracts the setter pattern from the line comment.
+// If the the line comment doesn't contain MutationCommentIdentifier
+// prefix, then it returns an empty string.
+func extractMutationPattern(lineComment string) string {
+	if !strings.HasPrefix(lineComment, mutationCommentIdentifier) {
+		return ""
+	}
+	return strings.TrimSpace(strings.TrimPrefix(lineComment, mutationCommentIdentifier))
+}
+
+func (fw *fieldWalker) visitField(node *yaml.RNode, n string) error {
+	// Visit fields with comments.
+	if comment := node.YNode().LineComment; comment != "" {
+		fmt.Fprintln(os.Stderr, "Parsing field comment", comment)
+		// Check if comment is a mutation annotation.
+		mutationPattern := extractMutationPattern(comment)
+		if mutationPattern == "" {
+			// The comment is not a mutation annotation.
+			return nil
+		}
+		if !annotatemutations.HasRef(mutationPattern) {
+			// Mutation comment is tagged but no valid reference found.
+			return fmt.Errorf("apply mutation comment found with no valid reference to source path")
+		}
+
+		resourceRef, refPath := annotatemutations.CommentToReference(mutationPattern)
+
+		// replace the setter names in comment pattern with provided values
+		replacementValue, replacementToken := annotatemutations.CommentToTokenField(mutationPattern, fw.mutationCount)
+		fw.mutationCount++
+		if replacementValue != "" {
+			node.YNode().Value = replacementValue
+		}
+
+		fw.annotation = append(fw.annotation, annotatemutations.Mutation{SourceRef: resourceRef, SourcePath: refPath, TargetPath: "$." + n, Token: replacementToken})
+		fw.parentProcessor.results = append(fw.parentProcessor.results, framework.ResultItem{Message: fmt.Sprintf("Parsed mutation in resource %q field %q", fw.fileName, n), Severity: framework.Info})
+	}
+	return nil
+}
+
+// visitKeys recurses over a complex yaml hierarchy
+func (fw *fieldWalker) visitKeys(object *yaml.RNode, p string) error {
```

Comment:
It would be good and consistent if you move the api layer to annotatemutations package and use main.go for cli to api conversion and vice versa. Please refer for other functions like `apply-setters` for such pattern.

Reply:
refactoring...